### PR TITLE
Add scheduler concept guide

### PR DESCRIPTION
**Pull request by sftim**
This PR is heavily based on PR #10739 by @Violet-Guo

I've drawn on it to migrate other content into the new section, tidy up wording, and add glossary references that looked appropriate.

~@Violet-Guo: how do you like the look of this?~
~I'd be fine to have you rebase #10739 and then cherry-pick the commits I've added here; if you do that, I can close this PR.~
~It can be the other way round if you want, keeping this PR open.~

Overall, this set of changes feels to me like they are a valid and worthwhile improvement compared to what is there now. For that reason I recommend merging as-is (via some route, see above) even if there is scope for further enhancement - these docs look OK to me.
If there are typos or technical details that aren't correct, I would be keen to get those fixed first.

/priority important-longterm

### CODE REVIEW

**Comment by sftim** [comment]
I'd also like to build on this with content from PR #11262.

**Comment by sftim** [comment]
OK, I think at this point I'll put this forward for review.

File: content/en/docs/concepts/scheduling/kube-scheduler.md [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+is the default scheduler for Kubernetes and runs as part of the
+{{< glossary_tooltip text="control plane" term_id="control-plane" >}}.
+kube-scheduler is designed so that, if you want and need to, you can
+write your own scheduling component and use that instead.
+
+For every newly created pods or other unscheduled pods, kube-scheduler
+selects a optimal node for them to run on.  However, every container in
+pods has different requirements for resources and every pod also has
+different requirements. Therefore, existing nodes need to be filtered
+according to the specific scheduling algorithm and strategy.
```
Comment (bsalamat) [inline content/en/docs/concepts/scheduling/kube-scheduler.md]: Please replace `scheduling algorithm and strategy` with `scheduling requirements`. The scheduling algorithm is the same for most pods. It is just the pod's requirements that make the scheduling outcome different.

File: content/en/docs/concepts/scheduling/kube-scheduler.md [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+are called _feasible_ nodes. If none of the nodes are suitable, the pod
+remains unscheduled until the scheduler is able to place it.
+
+The scheduler finds feasible Nodes for a Pod and then runs a set of
+functions to score the feasible Nodes and picks a Node with the highest
+score among the feasible ones to run the Pod. The scheduler then notifies
+the API server about this decision in a process called _binding_.
+
+Factors that need taken into account for scheduling decisions include
+individual and collective resource requirements, QoS (Quality of Service)
```
Comment (bsalamat) [inline content/en/docs/concepts/scheduling/kube-scheduler.md]: Please remove `QoS` and `deadlines` from this list. kube-scheduler does not take them into account.

File: content/en/docs/concepts/scheduling/kube-scheduler.md [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+requirements, hardware / software / policy constraints, affinity and
+anti-affinity specifications, data locality, inter-workload interference,
+deadlines, and so on.
+
+## Scheduling with kube-scheduler {#kube-scheduler-implementation}
+
+kube-scheduler selects a node for the pod in a 2-step operation:
+
+1. **Predicates** processing: filter the nodes
```
Comment (bsalamat) [inline content/en/docs/concepts/scheduling/kube-scheduler.md]: Let's remove `Predicates` and `Priorities` from here and replace them with `Filtering` and `Scoring`. We can add that these are called `predicates` and `priorities` in the internal logic of the scheduler, but I don't want these names to become official. In fact, we are changing them as a part of building the scheduling framework.

File: content/en/docs/concepts/scheduling/scheduler-perf-tuning.md [hunk 4]
```diff
@@ -10,13 +10,12 @@ weight: 70
...
 {{< feature-state for_k8s_version="1.14" state="beta" >}}
 
-Kube-scheduler is the Kubernetes default scheduler. It is responsible for
-placement of Pods on Nodes in a cluster. Nodes in a cluster that meet the
-scheduling requirements of a Pod are called "feasible" Nodes for the Pod. The
-scheduler finds feasible Nodes for a Pod and then runs a set of functions to
-score the feasible Nodes and picks a Node with the highest score among the
-feasible ones to run the Pod. The scheduler then notifies the API server about
-this decision in a process called "Binding".
+[kube-scheduler](/docs/concepts/scheduling/kube-scheduler/#kube-scheduler)
```
Comment (bsalamat) [inline content/en/docs/concepts/scheduling/scheduler-perf-tuning.md]: Please add this link, but leave the rest of text in this file. It wouldn't hurt to have a bit more information about the scheduler here.

File: content/en/docs/concepts/scheduling/kube-scheduler.md [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+   matches the Node's {{< glossary_tooltip text="label(s)" term_id="label" >}}.
+
+- `NoVolumeZoneConflict`: Evaluate if the {{< glossary_tooltip text="Volumes" term_id="volume" >}}
+  that a Pod requests are available on the Node, given the failure zone restrictions for
+  that storage.
+
+- `NoDiskConflict`: Evaluates if a Pod can fit on a Node due to the volumes it requests,
+   and those that are already mounted.
+
+- `MaxAzureDiskVolumeCount`: Ensures that the number of attached Azure Disk volumes
```
Comment (bsalamat) [inline content/en/docs/concepts/scheduling/kube-scheduler.md]: Please remove `MaxAzureDiskVolumeCount`, `MaxEBSVolumeCountPred`, and `MaxGCEPDVolumeCount`. These are deprecated in favor of `MaxCSIVolumeCount`.

File: content/en/docs/concepts/scheduling/kube-scheduler.md [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+- `MaxEBSVolumeCount`: Ensures that the number of attached ElasticBlockStore volumes
+  does not exceed a maximum value. This predicate only applies for AWS clusters.
+
+- `MaxGCEPDVolumeCount`: Ensures that the number of attached GCE PersistentDisk
+  volumes does not exceed a maximum value. This predicate only applies for GCE clusters.
+
+- `MaxCSIVolumeCount`: Decides how many {{< glossary_tooltip text="CSI" term_id="csi" >}}
+  volumes should be attached, and whether that's over a configured limit.
+
+- `GeneralPred`: Checks whether nonCriticalPredicates and essentialPredicates pass.
```
Comment (bsalamat) [inline content/en/docs/concepts/scheduling/kube-scheduler.md]: Please remove this. `GeneralPred` is a library function that runs a number of other predicates. It is not an independent predicate.

File: content/en/docs/concepts/scheduling/kube-scheduler.md [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+  can tolerate the Node's {{< glossary_tooltip text="taints" term_id="taint" >}}.
+
+- `CheckVolumeBinding`: Evaluates if a Pod can fit due to the volumes it requests.
+  This applies for both bound and unbound
+  {{< glossary_tooltip text="PVCs" term_id="persistent-volume-claim" >}}
+
+### Priorites Policies
+
+- `SelectorSpreadPriority`: Spreads Pods across hosts, considering Pods that
+   belonging to the same {{< glossary_tooltip text="PVCs" term_id="service" >}},
```
Comment (bsalamat) [inline content/en/docs/concepts/scheduling/kube-scheduler.md]: s/PVCs/service/

File: content/en/docs/concepts/scheduling/kube-scheduler.md (hunk 3 above)
Comment (sftim) [reply content/en/docs/concepts/scheduling/kube-scheduler.md]: Good to know, thanks.

File: content/en/docs/concepts/scheduling/kube-scheduler.md (hunk 5 above)
Comment (sftim) [reply content/en/docs/concepts/scheduling/kube-scheduler.md]: Noted

**Comment by sftim** [comment]
Revised & rebased

File: content/en/docs/concepts/scheduling/kube-scheduler.md (hunk 3 above)
Comment (sftim) [reply content/en/docs/concepts/scheduling/kube-scheduler.md]: Done

**Comment by sftim** [comment]
Hi @bsalamat 
I would like _not_ to further squash the commits in this PR, because:

1. Prow will squash this before merging the squash (as a fast-forward). I favor leaving the commits separate _in the PR_ though because this captures intent, and leaves that record there in case anyone ever wants to understand the motivation behind part of these changes.
2. This PR has multiple authors and I want to preserve a record of attribution.

**Comment by sftim** [comment]
@Violet-Guo if you have an opinion on squashing vs. not here, I'm keen to bear that in mind.

**Comment by Violet-Guo** [comment]
> if you have an opinion on squashing vs. not here, I'm keen to bear that in mind.

Sorry about the late response, I am agree with you.

**Comment by sftim** [comment]
Is this OK to merge unsquashed? (I & @Violet-Guo, the original author, think it is).

**Comment by sftim** [comment]
This PR might need revising as https://kubernetes.io/docs/concepts/configuration/scheduling-framework/ has been added since the text was originally written.

**Comment by Violet-Guo** [comment]
> Is this OK to merge unsquashed? (I & @Violet-Guo, the original author, think it is).
It's OK to merge it.

**Comment by bradtopol** [comment]
Great job!!!
/lgtm
/approve

**Pull request merged**